File: android/app/src/main/kotlin/io/devopen/subzilla/widget/NextDueDateHelper.kt
package io.devopen.subzilla.widget

import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

object NextDueDateHelper {
    private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS")

    fun parseDate(isoString: String): LocalDate? = try {
        LocalDate.parse(isoString, formatter)
    } catch (e: Exception) {
        // Fallback for full ISO with timezone
        try { LocalDate.parse(isoString.take(10)) } catch (e2: Exception) { null }
    }

    fun nextDueDate(startDate: LocalDate, frequency: String, today: LocalDate = LocalDate.now()): LocalDate {
        if (startDate.isAfter(today)) return startDate
        var candidate = when (frequency) {
            "daily" -> {
                val days = ChronoUnit.DAYS.between(startDate, today) + 1
                startDate.plusDays(days)
            }
            "weekly" -> {
                val weeks = ChronoUnit.WEEKS.between(startDate, today) + 1
                startDate.plusWeeks(weeks)
            }
            "monthly" -> {
                val months = ChronoUnit.MONTHS.between(startDate, today) + 1
                startDate.plusMonths(months)
            }
            "yearly" -> {
                val years = ChronoUnit.YEARS.between(startDate, today) + 1
                startDate.plusYears(years)
            }
            else -> startDate.plusMonths(1)
        }
        // Guard: if still <= today (edge case with month-end clamping), advance one more period
        if (!candidate.isAfter(today)) {
            candidate = when (frequency) {
                "daily" -> candidate.plusDays(1)
                "weekly" -> candidate.plusWeeks(1)
                "monthly" -> candidate.plusMonths(1)
                "yearly" -> candidate.plusYears(1)
                else -> candidate.plusMonths(1)
            }
        }
        return candidate
    }

    fun daysUntil(date: LocalDate, today: LocalDate = LocalDate.now()): Long =
        ChronoUnit.DAYS.between(today, date)
}

File: ios/SubsWidget/NextDueDateHelper.swift
import Foundation

enum SubFrequency: String {
    case daily, weekly, monthly, yearly
}

struct NextDueDateHelper {
    static func nextDueDate(startDate: Date, frequency: SubFrequency, from today: Date = Date()) -> Date {
        guard startDate > today else {
            let cal = Calendar.current
            switch frequency {
            case .daily:
                let days = cal.dateComponents([.day], from: startDate, to: today).day! + 1
                return cal.date(byAdding: .day, value: days, to: startDate)!
            case .weekly:
                let weeks = cal.dateComponents([.weekOfYear], from: startDate, to: today).weekOfYear! + 1
                return cal.date(byAdding: .weekOfYear, value: weeks, to: startDate)!
            case .monthly:
                let months = cal.dateComponents([.month], from: startDate, to: today).month! + 1
                let candidate = cal.date(byAdding: .month, value: months, to: startDate)!
                // candidate might still be <= today due to month-end clamping; advance one more if needed
                if candidate <= today {
                    return cal.date(byAdding: .month, value: 1, to: candidate)!
                }
                return candidate
            case .yearly:
                let years = cal.dateComponents([.year], from: startDate, to: today).year! + 1
                let candidate = cal.date(byAdding: .year, value: years, to: startDate)!
                if candidate <= today {
                    return cal.date(byAdding: .year, value: 1, to: candidate)!
                }
                return candidate
            }
        }
        return startDate
    }

    static func daysUntil(_ date: Date, from today: Date = Date()) -> Int {
        let cal = Calendar.current
        let fromDay = cal.startOfDay(for: today)
        let toDay = cal.startOfDay(for: date)
        return cal.dateComponents([.day], from: fromDay, to: toDay).day ?? 0
    }
}
